### PR TITLE
DruidInputSource can add new dimensions during re-ingestion

### LINKED ISSUE #9593: Can not add a column when re-indexing a datasource using DruidInputSource
### Affected Version

0.18

### Description

See the following ingestionSpec that can be used in Integration tests. The transformed column "newPage" does not show up in the dimensions for the new re-indexed dataSource. To work around this, I fall back to use the ingestSegmentFirehose instead
```
{
    "type": "index",
    "spec": {
        "ioConfig": {
            "type": "index",
            "inputSource": {
                "type": "druid",
                "dataSource": "%%DATASOURCE%%",
                "interval": "2013-08-31/2013-09-01"
            }
        },
        "tuningConfig": {
            "type": "index",
            "partitionsSpec": {
                "type": "dynamic"
            }
        },
        "dataSchema": {
            "dataSource": "%%REINDEX_DATASOURCE%%",
            "granularitySpec": {
                "type": "uniform",
                "queryGranularity": "SECOND",
                "segmentGranularity": "DAY"
            },
            "timestampSpec": {
                "column": "__time",
                "format": "iso"
            },
            "dimensionsSpec": {
                "dimensions": [
                    {"type": "string", "name": "language", "createBitmapIndex": false},
                    "user",
                    "unpatrolled",
                    "page",
                    "newPage",
                    "anonymous",
                    "namespace",
                    "country",
                    "region",
                    "city"
                ],
                "dimensionExclusions" : ["robot", "continent"]
            },
            "transformSpec": {
                "transforms": [
                    {
                        "type": "expression",
                        "name": "newPage",
                        "expression": "\"page\""
                    },
                    {
                        "type": "expression",
                        "name": "one-plus-triple-added",
                        "expression": "\"triple-added\" + 1"
                    },
                    {
                        "type": "expression",
                        "name": "double-deleted",
                        "expression": "deleted * 2"
                    }
                ]
            },
            "metricsSpec": [
                {
                    "type": "doubleSum",
                    "name": "added",
                    "fieldName": "added"
                },
                {
                    "type": "doubleSum",
                    "name": "triple-added",
                    "fieldName": "triple-added"
                },
                {
                    "type": "doubleSum",
                    "name": "one-plus-triple-added",
                    "fieldName": "one-plus-triple-added"
                },
                {
                    "type": "doubleSum",
                    "name": "deleted",
                    "fieldName": "deleted"
                },
                {
                    "type": "doubleSum",
                    "name": "double-deleted",
                    "fieldName": "double-deleted"
                },
                {
                    "type": "doubleSum",
                    "name": "delta",
                    "fieldName": "delta"
                }
            ]
        }
    }
}
```

### PR DESCRIPTION
**Pull request by suneet-s**
Fixes #9593 #9592 

### Description

Add integration tests for using transformSpec during ingestion. The tests test transformSpecs that add metrics and dimensions using both a Druid InputSource and the deprecated ingestSegmentFirehose.

Writing these tests surfaced #9589 #9591 When these issues are fixed, those tests can be re-enabled.

### CODE REVIEW

File: integration-tests/src/test/resources/indexer/wikipedia_index_queries_with_transform.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            "having":{"type":"greaterThan", "aggregation":"added_count_times_ten", "value":9000},
+            "intervals":[
+                "2013-08-31T00:00/2013-09-01T00:00"
+            ]
+        },
+        "expectedResults":[ {
+            "version" : "v1",
+            "timestamp" : "2013-08-31T00:00:00.000Z",
+            "event" : {
+                "added_count_times_ten" : 9050.0,
```
Comment (suneet-s) [inline integration-tests/src/test/resources/indexer/wikipedia_index_queries_with_transform.json]: typo in expected results
```suggestion
                "added_count_times_ten" : 27150.0,
```

File: integration-tests/src/test/resources/indexer/wikipedia_index_queries_with_transform.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
+[
+    {
+        "description":"having spec on post aggregation",
+        "query":{
+            "queryType":"groupBy",
+            "dataSource":"%%DATASOURCE%%",
+            "granularity":"day",
+            "dimensions":[
+                "page"
```
Comment (suneet-s) [inline integration-tests/src/test/resources/indexer/wikipedia_index_queries_with_transform.json]: Using a filter on "newPage" fails the test locally. But if I apply the transformation from another datasource it appears to work. Need to verify if this is actually a bug or not

**Comment by maytasm3** [comment]
Can you parameterize all the query and index json files? I think it's much harder to keep track of many many many different query and index json files we have for IT. Especially when they are 80-90% the same.

**Comment by suneet-s** [comment]
> Can you parameterize all the query and index json files? I think it's much harder to keep track of many many many different query and index json files we have for IT. Especially when they are 80-90% the same.

Across all `AbstractITBatchIndexTest` or within the new tests I've added in `ITTransformTest`?

It's hard to parameterize `ITTransformTest` because it's broken for most combinations. And there are 2 categories of tests. I think properly parameterizing these tests requires a re-think of what the test environment should look like because all the re-indexing tests start by indexing a datasource which slows down the test execution.

**Comment by maytasm3** [comment]
> > Can you parameterize all the query and index json files? I think it's much harder to keep track of many many many different query and index json files we have for IT. Especially when they are 80-90% the same.
> 
> Across all `AbstractITBatchIndexTest` or within the new tests I've added in `ITTransformTest`?
> 
> It's hard to parameterize `ITTransformTest` because it's broken for most combinations. And there are 2 categories of tests. I think properly parameterizing these tests requires a re-think of what the test environment should look like because all the re-indexing tests start by indexing a datasource which slows down the test execution.

I am thinking of parameterizing the json files themself across the new json files you added

**Comment by suneet-s** [comment]
> 
> I am thinking of parameterizing the json files themself across the new json files you added

I think parameterizing the json files is a great idea, but there's a lot of tight coupling between the json files and the base IT classes. I'm not sure where the best place to do this is because of that. 

I'd rather do the parameterization in a separate change - when #9591 #9592  or #9589  are fixed

File: integration-tests/src/test/resources/indexer/wikipedia_index_queries_with_transform.json (hunk 2 above)
Comment (suneet-s) [reply integration-tests/src/test/resources/indexer/wikipedia_index_queries_with_transform.json]: This is a bug filed #9593 for this

**Pull request merged**